src/gpu/nodeKernels.cu

```diff
@@ -223,41 +334,97 @@ __global__ void kFilterGroundPoints(size_t pointCount, const Vec3f sensor_up_vec
 	outNonGround[tid] = normalUpAngle > ground_angle_threshold;
 }
 
-__global__ void kProcessBeamSamplesFirstLast(size_t beamCount, int samplesPerBeam, MultiReturnPointers beamSamples,
-                                             MultiReturnPointers first, MultiReturnPointers last, const Mat3x4f* beamsWorld)
+__global__ void kReduceDivergentBeams(size_t beamCount, int samplesPerBeam, rgl_return_mode_t returnMode,
+                                      const RaytraceRequestContext* ctx)
 {
 	LIMIT(beamCount);
 
 	const auto beamIdx = tid;
-	int firstIdx = -1;
-	int lastIdx = -1;
-	for (int sampleIdx = 0; sampleIdx < samplesPerBeam; ++sampleIdx) {
-		if (beamSamples.isHit[beamIdx * samplesPerBeam + sampleIdx] == 0) {
+	const auto firstSampleInBeamIdx = beamIdx * samplesPerBeam;
+	int first = -1, second = -1, last = -1;
+	int strongest = -1, secondStrongest = -1;
+	int sampleIdx = firstSampleInBeamIdx;
+
+	for (; sampleIdx < firstSampleInBeamIdx + samplesPerBeam; ++sampleIdx) {
+		if (ctx->mrSamples.isHit[sampleIdx] == 1) {
+			first = sampleIdx;
+			second = first;
+			last = first;
+			strongest = sampleIdx;
+			secondStrongest = strongest;
+			break;
+		}
+	}
+
+	const auto returnCount = returnMode >> RGL_RETURN_MODE_BIT_SHIFT;
+
+	// There were no hits within beam samples.
+	if (first == -1) {
+		for (int returnIdx = 0; returnIdx < returnCount; ++returnIdx) {
+			const auto returnPointIdx = beamIdx * returnCount + returnIdx;
+			const auto returnType = (returnMode >> (returnIdx * RGL_RETURN_TYPE_BIT_SHIFT)) & 0xff;
+			saveReturnAsNonHit(ctx, beamIdx, returnPointIdx, returnType);
+		}
+		return;
+	}
+
+	// Base initial values on first sample that returned a hit.
+	for (sampleIdx = sampleIdx + 1; sampleIdx < firstSampleInBeamIdx + samplesPerBeam; ++sampleIdx) {
+		if (ctx->mrSamples.isHit[sampleIdx] == 0) {
 			continue;
 		}
-		auto currentFirstDistance = firstIdx >= 0 ? beamSamples.distance[beamIdx * samplesPerBeam + firstIdx] : FLT_MAX;
-		auto currentLastDistance = lastIdx >= 0 ? beamSamples.distance[beamIdx * samplesPerBeam + lastIdx] : -FLT_MAX;
-		if (beamSamples.distance[beamIdx * samplesPerBeam + sampleIdx] < currentFirstDistance) {
-			firstIdx = sampleIdx;
+
+		const auto currentSampleDistance = ctx->mrSamples.distance[sampleIdx];
```

Comment:
`currentSampleDistance` -> `currentDistance` for consistency with `currentIntensity` defined later.